### PR TITLE
Docs: Checkout

### PR DESCRIPTION
**Pull request by Fifciu**
### Related Issues

closes #

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature
- [ ] (Next only) I've followed this [instruction](https://docs-next.vuestorefront.io/contributing/creating-changelog.html) and I've created a `.js` file with information about my Pull Request

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: packages/core/docs/guide/checkout.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Checkout
+Checkout is a process of providing information about shipping and billing details, and picking certain shipping and payment method. Then with all the data, we are able to place an order and pay for it.
+
+## Shipping
+To access shipping details, we can use property of `useShipping` called `shipping`.
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:5]: maybe its worth explaining very briefly what are shipping details

File: packages/core/docs/guide/checkout.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
+# Checkout
+Checkout is a process of providing information about shipping and billing details, and picking certain shipping and payment method. Then with all the data, we are able to place an order and pay for it.
+
+## Shipping
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:4]: as @NataliaTepluhina was suggesting let's have more descriptive subtitles (also for SEO)

File: packages/core/docs/guide/checkout.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      await load();
+    });
+
+    return {
+      shipping
+    };
+  }
+}
+```
+`shipping` property will return `null` if the address is not saved on the server or just has not been loaded. In order to save shipping details, you could use `save` method:
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:27]:
```suggestion
`shipping` property will return `null` if the address is not saved on the server or just has not been loaded. In order to save shipping details, you can use `save` method:
```

also it's worth explaining what `save` means a bit further (eg it will be persisted in a composable state / available each time you call `load` or similar )

File: packages/core/docs/guide/checkout.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    return {
+      shippingForm,
+      save
+    };
+  }
+}
+</script>
+```
+
+## Shipping providers
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:59]: The below paragraph will probably be unable to understand by anyone who doesn't already know what shipping provider component is. The explanation is missing essential information like why we need this component, what is a shipping provider component, what are responsibilities of such component etc.

Comment (filrak) [reply packages/core/docs/guide/checkout.md:59]: also what is a proper place, please avoid such constructions in the docs - be specific

File: packages/core/docs/guide/checkout.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+We delegate whole logic of selecting shipping method to the dedicated component called **VsfShippingProvider.vue**. Each shipping extension shares one. All we have to do is import and put it in a proper place.
+```vue
+<VsfShippingProvider
+  @submit="$router.push('/checkout/billing')"
+/>
+```
+
+VsfShippingProvider emits `submit` event when it is ready to move to the next step.
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:67]:
```suggestion
VsfShippingProvider emits `submit` event when it is ready to move to the next step.
```
how can component be ready to move? :D

File: packages/core/docs/guide/checkout.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+Component shares 5 hooks with async support:
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:69]:
```suggestion
You can pass asynchronous functions as `VsfShippingProvider` props to hook into different events within it's lifecycle and override initial function parameters or react to specific events like method selection
```

File: packages/core/docs/guide/checkout.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+- **beforeLoad** - called before loading shipping methods
+- **afterLoad** - called after loading shipping methods
+- **onSelected** - called after selecting shipping method
+- **onSelectedDetailsChanged** - called after modifying currently picked shipping method, e.g. selecting parcel locker on the map
+- **onError** - called when some operation throws an error
+
+We can inject in the hook via props:
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:76]: we can just have a code with comments + with functions in a script section (or their types below) instead of a list AND code. 

You will avoid repetition by doing sth like this:

```
<VsfShippingProvider
  <!-- called before loading shipping methods -->
  :beforeLoad="(config) => config"
  <!-- ... -->
  :afterLoad="..."
  :onSelected="..."
  :onSelectedDetailsChanged="..."
  :onError="..."
/>

```

File: packages/core/docs/guide/checkout.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+  :beforeLoad="beforeLoad"
+  :afterLoad="afterLoad"
+  :onSelected="onSelected"
+  :onSelectedDetailsChanged="onSelectedDetailsChanged"
+  :onError="onError"
+/>
+```
+
+
+## Billing
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:88]: all comments from above apply here

**Review by filrak — CHANGES_REQUESTED** [review]

File: packages/core/docs/guide/checkout.md (hunk 7 above)
Comment (Fifciu) [reply packages/core/docs/guide/checkout.md:82]: It isn't a proper markup, so the colors under the line with the comment would be not proper too.
image

File: packages/core/docs/guide/checkout.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    };
+  }
+}
+</script>
+```
+
+Shipping details stored on the server with `save` method will be possible to load after a refresh with `load` method.
+
+## Shipping providers
+Shipping provider is an aggregator that provides us one or more shipping methods.
```
Comment (andrzejewsky) [inline packages/core/docs/guide/checkout.md:66]: please mention that's integration, not only the "aggregator", always one provider components mean one 3rd party provider of either payments or shippings

File: packages/core/docs/guide/checkout.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+`VsfShippingProvider` emits `submit` event when shipping method is properly picked, configured and user clicks `Continue to billing` button.
+
+### Hooks
+You can pass asynchronous functions as `VsfShippingProvider` props to hook into different events within it's lifecycle and override initial function parameters or react to specific events like method selection
+- **beforeLoad** - called before loading shipping methods
+- **afterLoad** - called after loading shipping methods
+- **onSelected** - called after selecting shipping method
+- **onSelectedDetailsChanged** - called after modifying currently picked shipping method, e.g. selecting parcel locker on the map
+- **onError** - called when some operation throws an error
```
Comment (andrzejewsky) [inline packages/core/docs/guide/checkout.md:89]: mention signature: what arguments I have and what values they are returning

File: packages/core/docs/guide/checkout.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+      save
+    };
+  }
+}
+</script>
+```
+Billing details stored on the server with `save` method will be possible to load after a refresh with `load` method.
+
+## Payment providers
+Payment provider is an aggregator that provides us one or more payment methods.
```
Comment (andrzejewsky) [inline packages/core/docs/guide/checkout.md:164]: same here

File: packages/core/docs/guide/checkout.md [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+`VsfPaymentProvider` emits `status` event with boolean in the payload which tells whether everything is ready for call `pay` or not. Use this information to enable/disable `Place order` button.
+
+### Hooks 
+You can pass asynchronous functions as `VsfPaymentProvider` props to hook into different events within it's lifecycle and override initial function parameters or react to specific events like method selection
+- **beforeLoad** - called before loading payment methods
+- **afterLoad** - called after loading payment methods
+- **onSelected** - called after selecting payment method
+- **onSelectedDetailsChanged** - called after modifying currently picked payment method, e.g. changing credit card's details
+- **onError** - called when some operation throws an error
```
Comment (andrzejewsky) [inline packages/core/docs/guide/checkout.md:221]: and here

**Review by andrzejewsky — CHANGES_REQUESTED** [review]

**Review by filrak — COMMENTED** [review]
Also you use "we" and "you" - decide on one

File: packages/core/docs/guide/checkout.md [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+      await load();
+    });
+
+    return {
+      shipping
+    };
+  }
+}
+```
+`shipping` property will return `null` if the address is not saved on the server or has not been loaded.
```
[the remark this answers is not in the document]
Comment (Fifciu) [reply packages/core/docs/guide/checkout.md:29]: `or` fits pretty well there in my opinion.

I could rephrase it to smth like that at first (`or` still present):
`shipping` property returns `null` if the `load` function was not invoked or the `save` function was not invoked before

File: packages/core/docs/guide/checkout.md [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+
+    return {
+      billingForm,
+      save
+    };
+  }
+}
+</script>
+```
+Billing details stored on the server with `save` method will be possible to load after a refresh with `load` method.
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md]:
```suggestion

```

File: packages/core/docs/guide/checkout.md [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+
+## Order
+After providing every crucial information by the user, we are ready to *make an order*. To do that, we have to call a `make` method from the `useMakeOrder` composable.
```
[the remark this answers is not in the document]
Comment (filipsobol) [reply packages/core/docs/guide/checkout.md]:
```suggestion
Once the user provides all required information, we can *make an order* by calling the `make` method in `useMakeOrder` composable.
```

**Comment by filrak** [comment]
Also please correct all other documents that either list composables or describe the ones that were changed

File: packages/core/docs/guide/checkout.md [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+    return {
+      make
+    }
+  }
+}
+```
+
+It creates an order but we need to perform additional actions:
+- redirect to thank you page
+- clear a cart
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md:180]:
```suggestion
When the order is created, we can redirect the user to the page thanking them for making an order and refresh the cart.
```

`It creates an order but we need to perform additional actions:` sounds like there are some other steps required to finalize the order.

File: packages/core/docs/guide/checkout.md [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+      await make();
+      context.root.$router.push(`/checkout/thank-you?order=${order.value.id}`);
+      setCart(null);
+    }
+  }
+}
+```
+
+## Payment providers
+Payment provider is an aggregator that provides us one or more payment methods. It is also an integration, one provider component always means one 3rd party provider of payments.
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md]:
```suggestion
A payment provider is a component that provides one or more payment methods. One such component integrates one 3rd party provider of payments.
```

File: packages/core/docs/guide/checkout.md [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+      setCart(null);
+    }
+  }
+}
+```
+
+## Payment providers
+Payment provider is an aggregator that provides us one or more payment methods. It is also an integration, one provider component always means one 3rd party provider of payments.       
+
+To give you the best developer experience, we delegate whole logic of selecting and configuring payment method to the dedicated component called **VsfPaymentProvider.vue**. It takes care of:
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md]:
```suggestion
For the best developer experience, we delegate the whole logic of selecting and configuring payment methods to the dedicated component named **VsfPaymentProvider.vue**. It takes care of:
```

File: packages/core/docs/guide/checkout.md [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+- Loading and showing available payment methods
+- Picking and configuring payment method
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md]: List punctuation

File: packages/core/docs/guide/checkout.md [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+
+The first thing we have to do is import and put the component inside `pages/Payment.vue` as a [last part of the Checkout](https://github.com/vuestorefront/vue-storefront/blob/next/packages/commercetools/theme/pages/Checkout/Payment.vue#L108):
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md]:
```suggestion
All we have to do is to use it in the `pages/Checkout/Payment.vue` as a [last part of the Checkout](https://github.com/vuestorefront/vue-storefront/blob/next/packages/commercetools/theme/pages/Checkout/Payment.vue#L108):
```

File: packages/core/docs/guide/checkout.md [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+
+export default {
+  components: {
+    VsfPaymentProvider
+  }
+}
+</script>
+```
+
+Then we have to support making a payment - each package with a payment provider might force us to use a bit different approach. We found that there will be 2 most common ones:
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md:225]:
```suggestion
The next step is making a payment. Each package with a payment provider might use a slightly different approach, but below we described are two most common.
```

File: packages/core/docs/guide/checkout.md [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+      await load();
+    });
+
+    return {
+      shipping
+    };
+  }
+}
+```
+`shipping` property returns `null` if the `load` function was not invoked or nothing is saved.
```
[the remark this answers is not in the document]
Comment (Fifciu) [reply packages/core/docs/guide/checkout.md:29]: We cannot assume it is always saved on the server. Also @filrak wanted to make it clear that we need to call `load` (in his CR above)

File: packages/core/docs/guide/checkout.md [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+
+    return {
+      shipping
+    };
+  }
+}
+```
+`shipping` property returns `null` if the `load` function was not invoked or nothing is saved.    
+
+You can use `save` method, in order to save shipping details.
```
[the remark this answers is not in the document]
Comment (Fifciu) [reply packages/core/docs/guide/checkout.md]: Fair enough

File: packages/core/docs/guide/checkout.md [hunk 24]
```diff
@@ -293,19 +294,19 @@ export default {
...
-- **beforePay** `(paymentDetails => paymentDetails)` - called before pay
-- **afterPay** `(paymentResponse => void)` - called after pay
-- **beforeSelectedDetailsChange** `(details => details)` - called before modifying currently picked payment method, e.g. changing credit card's details
-- **afterSelectedDetailsChange** `(details => void)` - called after modifying currently picked payment method, e.g. changing credit card's details
-- **onError** `(({ action, error }) => void)` - called when some operation throws an error
+We can pass asynchronous functions to the `VsfPaymentProvider` component to hook into different events within its lifecycle, override initial function parameters or react to specific events like method selection. We will call these methods "hooks".
+
+Because every payment provider is different, not all of them are present in every integration. Always refer to the documentation of a specific provider to learn which hooks are available.
+
+- **beforeLoad** `(config => config)` - Called before loading payment methods/
```
Comment (filipsobol) [inline packages/core/docs/guide/checkout.md:301]:
```suggestion
- **beforeLoad** `(config => config)` - Called before loading payment methods.
```

File: packages/core/docs/commercetools/composables/use-make-order.md [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+
+- `make` - function for making an order. This method accepts a single optional `params` object. The `params` has the following option:
+ 
+    - `customQuery?: CustomQuery`
+
+```ts
+type CustomQuery = Record<string, string>
+```
+
+- `order: Order` - a main data object that contains a made order.
```
Comment (filrak) [inline packages/core/docs/commercetools/composables/use-make-order.md:17]: please paste order object here if possible

File: packages/core/docs/guide/checkout.md [hunk 26]
```diff
@@ -0,0 +1,324 @@
...
+  :beforePay="beforePay"
+  :afterPay="afterPay"
+  :beforeSelectedDetailsChange="beforeSelectedDetailsChange"
+  :afterSelectedDetailsChange="afterSelectedDetailsChange"
+  :onError="onError"
+/>
+```
+
+### Why some integrations have a mocked VsfPaymentProvider?
+There are eCommerce backends which do not provide any payment methods out-of-the-box, e.g. commercetools. For these, we provide mocked component to let user go through the whole checkout. We are using external providers with dedicated VsfPaymentProvider for them. One example is a [Checkout.com integration](https://github.com/vuestorefront/checkout-com).
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:324]:
```suggestion
There are eCommerce backends which do not provide any payment methods out-of-the-box, e.g. commercetools. For these, we provide mock component that has exactly same interface as a real integration so you can easily swap it with a payment integration of your choice. You can find available payment integrations [here](/v2/integrations/)
```

File: packages/core/docs/guide/checkout.md [hunk 27]
```diff
@@ -0,0 +1,324 @@
...
+    return {
+      billingForm,
+      save
+    };
+  }
+}
+</script>
+```
+
+## Making an order
```
Comment (filrak) [inline packages/core/docs/guide/checkout.md:163]: I know in many cases the order is opposite but maybe we should have this section at the end?

**Review by filrak — CHANGES_REQUESTED** [review]
can you add similar section at the end? https://docs.vuestorefront.io/v2/guide/user-profile.html#integration-references

File: packages/core/docs/guide/checkout.md (hunk 27 above)
Comment (Fifciu) [reply packages/core/docs/guide/checkout.md:163]: I am not sure about that. In PaymentProvider part, I mention about doing payment after order and using `useMakeOrder` which requires to put Order part before.
